common4j/src/main/com/microsoft/identity/common/java/opentelemetry/CertBasedAuthTelemetryHelper.java

```diff
@@ -0,0 +1,152 @@
+                    AttributeName.cert_based_auth_challenge_handler.name(),
+                    challengeHandlerName);
+        }
+    }
+
+    /**
+     * Sets attribute that indicates if a PivProvider instance is already present in the
+     *  Java Security static list upon adding a new instance.
+     * @param present true if PivProvider instance present; false otherwise.
+     */
+    public static void setExistingPivProviderPresent(@NonNull final boolean present) {
+        if (isCorrectSpan()) {
+            final Span span = Span.current();
+            span.setAttribute(
+                    AttributeName.cert_based_auth_existing_piv_provider_present.name(),
+                    present);
+        }
+    }
+
+    /**
+     * Indicates on the Span that CBA was successful and then ends current Span.
+     */
+    public static void setResultSuccess() {
+        if (isCorrectSpan()) {
+            final Span span = Span.current();
+            span.setStatus(StatusCode.OK);
+            span.end();
+        }
+    }
+
+    /**
+     * Indicates on the Span that CBA failed and then ends current Span.
+     * This method should mainly be used for cases without an exception,
+     *  such as user cancellation, no certificates on smartcard, etc.
+     * @param message descriptive cause of failure message.
+     */
+    public static void setResultFailure(@NonNull final String message) {
+        if (isCorrectSpan()) {
+            final Span span = Span.current();
+            //setting the error_message attribute manually since there's no exception to record.
+            span.setAttribute(
+                    "error_message",
+                    message);
+            span.setStatus(StatusCode.ERROR);
+            span.end();
+        }
+    }
+
+    /**
+     * Indicates on the Span that CBA failed and then ends current Span.
+     * @param exception exception thrown upon error.
+     */
+    public static void setResultFailure(@NonNull final Exception exception) {
+        if (isCorrectSpan()) {
+            final Span span = Span.current();
+            span.recordException(exception);
+            span.setStatus(StatusCode.ERROR);
+            span.end();
+        }
+    }
+
+    /**
+     * Sets attribute that indicates the user's choice for CBA (smartcard or on-device).
+     * This attribute is currently being set to an empty string, but it will be utilized
+     *  in future updates to the CBA dialogs.
+     * @param choice string indicating user's choice for CBA.
+     */
+    public static void setUserChoice(@NonNull final String choice) {
+        if (isCorrectSpan()) {
+            final Span span = Span.current();
+            span.setAttribute(
+                    AttributeName.cert_based_auth_user_choice.name(),
+                    choice);
+        }
+    }
+
+    /**
+     * Checks if the current Span is related to CBA.
+     * Used to ensure CBA attributes are not being set for other unrelated Spans.
```

Comment:
It would be really good to know how often the CBA code can run outside of the CBA span

Reply:
Can you enumerate the edge cases here?

Or if this has high likelihood of happening then we have a bigger problem in the code

Reply:
It has a pretty slim to none likelihood at the moment of happening (based on my latest testing). 
But I did include it because CBA can't be encapsulated by a single try/catch block, so the calls to the helper class may seem scattered to someone unfamiliar to the code. Also, the check does gives me some peace of mind by ensuring that CBA attributes will never ever be set for any other unrelated span.
But if this check does seem more confusing or unnecessary than helpful, I'm ok with leaving it out too.